### PR TITLE
Version bump to 4.0.0

### PR DESCRIPTION
**Pull request by kiskoza**
### Description

It looks like we're ready for a 4.0 release :tada: . We have a few breaking changes and a lot of performance and code style improvements! See the changelog for details

### Your checklist for this pull request

- [x] Link the changelog entries to the corresponding PRs when they have relevant conversations
- [x] Compare the final code performance with the previous version, e.g.: running #179 's script on both codes.
- [x] Look into #272 to make sure we're not introducing new problems

### CODE REVIEW

**Comment by kiskoza** [comment]
Running the scipt gave me these results on a Linux machine with Ruby 3.2.0

<details>
<summary>v3.4.1</summary>

```
adding Sheet 1
adding Sheet 2
adding Sheet 3
adding Sheet 4
adding Sheet 5
adding Sheet 6
adding Sheet 7
adding Sheet 8
adding Sheet 9
adding Sheet 10
RAM: 336.9MB used
================================================================================
finished adding data, now serializing to example.xlsx...
================================================================================
RAM: 345.4MB used
================================================================================
finished serialize, now trying GC...
================================================================================
RAM: 271.9MB used
================================================================================
```

</details>

<details>
<summary>v4.0.0</summary>

```
adding Sheet 1
adding Sheet 2
adding Sheet 3
adding Sheet 4
adding Sheet 5
adding Sheet 6
adding Sheet 7
adding Sheet 8
adding Sheet 9
adding Sheet 10
RAM: 211.5MB used
================================================================================
finished adding data, now serializing to example.xlsx...
================================================================================
RAM: 220.7MB used
================================================================================
finished serialize, now trying GC...
================================================================================
RAM: 161.5MB used
================================================================================
```

</details>

It went from 345 MB to 220 MB before the GC :tada:

**Comment by kiskoza** [comment]
Running the `test/profile_memory.rb` file gave me these results on a Linux machine with Ruby 3.2.0

<details>
<summary>v3.4.1</summary>

```
Total allocated: 446962203 bytes (6344981 objects)
Total retained:  352 bytes (5 objects)

allocated memory by gem
-----------------------------------
 414055363  caxlsx/lib
  32827242  other
     44830  htmlentities-4.3.4
     33688  lib
      1080  rubyzip-2.3.2
```

</details>

<details>
<summary>v4.0.0</summary>

```
Total allocated: 48001331 bytes (503447 objects)
Total retained:  871248 bytes (10174 objects)

allocated memory by gem
-----------------------------------
  47899771  caxlsx/lib
     44830  htmlentities-4.3.4
     33688  lib
     21962  other
      1080  rubyzip-2.3.2
```

</details>

**Comment by kiskoza** [comment]
Running the `test/profile.rb` file gave me these results on a Linux machine with Ruby 3.2.0

<details>
<summary>v3.4.1</summary>

```
Measure Mode: wall_time
Thread ID: 80
Fiber ID: 60
Total: 29.716476
Sort by: self_time
```

</details>

<details>
<summary>v4.0.0</summary>

```
Measure Mode: wall_time
Thread ID: 80
Fiber ID: 60
Total: 28.835403
Sort by: self_time
```

</details>

We're a bit better here, but the change is not significant. I've run this multiple times and v4 is always a bit faster

**Comment by kiskoza** [comment]
Running the `test/benchmark.rb` file gave me these results on a Linux machine with Ruby 3.2.0

<details>
<summary>v3.4.1</summary>

```
Rehearsal ------------------------------------------------------------------
axlsx_merged_cells               3.368247   0.028598   3.396845 (  3.444246)
axlsx_noautowidth                1.415493   0.004030   1.419523 (  1.440040)
axlsx_autowidth                  1.415674   0.003979   1.419653 (  1.421398)
axlsx_shared                     1.408692   0.016008   1.424700 (  1.425980)
axlsx_stream                     1.404997   0.008004   1.413001 (  1.413172)
axlsx_zip_command                1.241640   0.016021   1.347353 (  1.346054)
csv                              0.250811   0.003945   0.254756 (  0.264177)
-------------------------------------------------------- total: 10.675831sec

                                     user     system      total        real
axlsx_merged_cells               3.462212   0.003971   3.466183 (  3.467182)
axlsx_noautowidth                1.321460   0.003973   1.325433 (  1.340534)
axlsx_autowidth                  1.401756   0.004014   1.405770 (  1.415368)
axlsx_shared                     1.350669   0.000000   1.350669 (  1.369038)
axlsx_stream                     1.405090   0.003960   1.409050 (  1.429426)
axlsx_zip_command                1.228253   0.023971   1.337349 (  1.354161)
csv                              0.139051   0.003992   0.143043 (  0.167172)
```

</details>

<details>
<summary>v4.0.0</summary>

```
Rehearsal ------------------------------------------------------------------
axlsx_merged_cells               2.781255   0.004340   2.785595 (  2.786963)
axlsx_noautowidth                0.820946   0.004060   0.825006 (  0.830436)
axlsx_autowidth                  0.805101   0.004011   0.809112 (  0.822355)
axlsx_shared                     0.728441   0.007966   0.736407 (  0.745791)
axlsx_stream                     0.778496   0.004025   0.782521 (  0.793353)
axlsx_zip_command                0.754600   0.020026   0.876300 (  0.887524)
csv                              0.223400   0.016046   0.239446 (  0.323548)
--------------------------------------------------------- total: 7.054387sec

                                     user     system      total        real
axlsx_merged_cells               2.809398   0.011871   2.821269 (  2.859801)
axlsx_noautowidth                0.780830   0.000000   0.780830 (  0.791665)
axlsx_autowidth                  0.759333   0.000000   0.759333 (  0.769499)
axlsx_shared                     0.766077   0.007946   0.774023 (  0.784777)
axlsx_stream                     0.846095   0.003985   0.850080 (  0.861349)
axlsx_zip_command                0.700011   0.004035   0.794656 (  0.803049)
csv                              0.201997   0.019952   0.221949 (  0.233739)
```

</details>

On this benchmark, v4 looks a lot better :rocket:

**Comment by tagliala** [comment]
I can confirm benchmark results, v4 is ~ 60% faster on M1 Pro (x64 Ruby 3.2)

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
+- **October.XX.23**: 4.0.0
   - Drop support for Ruby versions < 2.6
-  - Added frozen string literals
-  - Fix `SimpleTypedList#to_a` and `SimpleTypedList#to_ary` returning the internal list instance
-  - Remove ability to set `u=` to true in favor of using :single or one of the other underline options
-  - Fix `Workbook#sheet_by_name` not returning sheets with encoded characters in the name
-  - Raise exception if `axlsx_styler` gem is present as its code was merged directly into `caxlsx` in v3.3.0
-  - Add 'SortState' and 'SortCondition' classes to the 'AutoFilter' class to add sorting to the generated file.
+  - [PR #219](https://github.com/caxlsx/caxlsx/pull/219) Added frozen string literals
+  - [PR #223](https://github.com/caxlsx/caxlsx/pull/223)Fix `SimpleTypedList#to_a` and `SimpleTypedList#to_ary` returning the internal list instance
```
Comment (tagliala) [inline CHANGELOG.md:9]: Missing whitespace 
```suggestion
  - [PR #223](https://github.com/caxlsx/caxlsx/pull/223) Fix `SimpleTypedList#to_a` and `SimpleTypedList#to_ary` returning the internal list instance
```

**Review by tagliala — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
+  - [PR #225](https://github.com/caxlsx/caxlsx/pull/225) Remove ability to set `u=` to true in favor of using :single or one of the other underline options
+  - [PR #239](https://github.com/caxlsx/caxlsx/pull/239) Fix `Workbook#sheet_by_name` not returning sheets with encoded characters in the name
+  - [PR #212](https://github.com/caxlsx/caxlsx/pull/212) Raise exception if `axlsx_styler` gem is present as its code was merged directly into `caxlsx` in v3.3.0
+  - [PR #286](https://github.com/caxlsx/caxlsx/pull/286) Add 'SortState' and 'SortCondition' classes to the 'AutoFilter' class to add sorting to the generated file.
   - [PR #189](https://github.com/caxlsx/caxlsx/pull/189) - Make `Axlsx::escape_formulas` true by default to mitigate [Formula Injection](https://www.owasp.org/index.php/CSV_Injection) vulnerabilities.
```
Comment (tagliala) [inline CHANGELOG.md:14]: would it be possible to highlight that this is a breaking change?

```suggestion
  - [PR #189](https://github.com/caxlsx/caxlsx/pull/189) - **Breaking**: Make `Axlsx::escape_formulas` true by default to mitigate [Formula Injection](https://www.owasp.org/index.php/CSV_Injection) vulnerabilities.
```

**Review by tagliala — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (kiskoza) [reply CHANGELOG.md]: Yeah, it makes sense. I marked a few others as breaking as well

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
+  - [PR #225](https://github.com/caxlsx/caxlsx/pull/225) **breaking** Remove ability to set `u=` to true in favor of using :single or one of the other underline options
+  - [PR #239](https://github.com/caxlsx/caxlsx/pull/239) Fix `Workbook#sheet_by_name` not returning sheets with encoded characters in the name
+  - [PR #212](https://github.com/caxlsx/caxlsx/pull/212) **breaking** Raise exception if `axlsx_styler` gem is present as its code was merged directly into `caxlsx` in v3.3.0
+  - [PR #286](https://github.com/caxlsx/caxlsx/pull/286) Add 'SortState' and 'SortCondition' classes to the 'AutoFilter' class to add sorting to the generated file.
+  - [PR #189](https://github.com/caxlsx/caxlsx/pull/189) **breaking** Make `Axlsx::escape_formulas` true by default to mitigate [Formula Injection](https://www.owasp.org/index.php/CSV_Injection) vulnerabilities.
+  - [PR #269](https://github.com/caxlsx/caxlsx/pull/269) Add optional interpolation points to icon sets
+  - [PR #304](https://github.com/caxlsx/caxlsx/pull/304) Fix data validations for none type validations
```
Comment (tagliala) [inline CHANGELOG.md:16]: sorry, another request: can we group breaking changes together at the beginning of the changelog?

**Review by tagliala — CHANGES_REQUESTED** [review]

**Review by tagliala — APPROVED** [review]

**Pull request merged**